attribute-projection-registry/build.gradle.kts

```diff
@@ -1,15 +1,42 @@
 plugins {
   `java-library`
   jacoco
-  id("org.hypertrace.jacoco-report-plugin")
-  id("org.hypertrace.publish-plugin")
+  alias(commonLibs.plugins.hypertrace.jacoco)
+  alias(commonLibs.plugins.hypertrace.publish)
+}
+
+configurations {
+  compileClasspath {
+    resolutionStrategy.activateDependencyLocking()
+  }
+  runtimeClasspath {
+    resolutionStrategy.activateDependencyLocking()
+  }
+}
+
+dependencyLocking {
+  lockMode.set(LockMode.STRICT)
+}
+
+tasks.register("resolveAndLockAll") {
+  notCompatibleWithConfigurationCache("Filters configurations at execution time")
+  doFirst {
+    require(gradle.startParameter.isWriteDependencyLocks) { "${path} must be run from the command line with the `--write-locks` flag" }
+  }
+  doLast {
+    configurations.filter {
+      // Add any custom filtering on the configurations to be resolved
+      it.isCanBeResolved
+    }.forEach { it.resolve() }
+  }
 }
 
 dependencies {
-  api(project(":attribute-service-api"))
-  implementation(project(":attribute-projection-functions"))
-  implementation("com.google.guava:guava:32.0.1-jre")
-  testImplementation("org.junit.jupiter:junit-jupiter:5.8.2")
+  api(projects.attributeServiceApi)
+  api(platform(commonLibs.hypertrace.bom))
```

Comment:
On the fence if plugin should auto add bom to each project. Seems like it would make things easier.

Reply:
If we are using version catalogs and doing dependency locking, what does the platform-plugin do additionally?

Reply:
Not sure if you read the link I sent out earlier, but it dives into this exact question. Including again here in case: https://docs.gradle.org/current/userguide/platforms.html#sub:platforms-vs-catalog , but let me see if I can do a better job with some examples. Apologies if this is repeating things that are already obvious, but I figure this will be a useful write up to share more broadly anyway so hopefully not wasted.

Let's take a project that previously declared:
```
implementation("com.google.guava:guava:32.0.1-jre")
```

A basic version catalog basically just gives type safe accessors. It just saves a user from typing that out which means the versions in that project will always be in sync.
```
implementation(libs.guava)
```

So far so good, but that's one repo. So now we can start publishing a version catalog, and we can do the same thing across repositories. I believe this is the experiment you were working with. But at this point, we've just kicked the problem down the road, because we have to declare the version of the catalog artifact itself that we want to use. Different repos may be using different catalogs. So that's where we introduced a dynamic version - always use the most recent version of catalog:
```
  versionCatalogs {
    create("commonLibs") {
      from("org.hypertrace.bom:hypertrace-version-catalog:+")
    }
  }
```

This introduces our next issue - repeatable builds. Having a build succeed today and then fail tomorrow because the version catalog changed with no source code changes sucks. So in come dependency locks. These basically materialize the resolution result and are checked in. There are equivalents in many languages (e.g. go, javascript/npm). So now, even if a new catalog is published, I have to opt in to pulling down that update.

Now what's the purpose of a dependency platform? It adds constraints on the resolution result even for dependencies you don't declare. So let's say we've got a second project, and this one has a dependency on grpc-api but is not directly using guava:
```
  implementation(libs.grpc.api) // "io.grpc:grpc-api:1.56.1"
```
Well grpc-api has a dependency on guava `"com.google.guava:guava:31.1-android"`, so this project would compile and run against that older version of guava - now we've got different projects compiling against different guava versions. When they're resolved together, conflict resolution has to select a single version for runtime (by default the newest one) - but that's a different version than the second project compiled against.

So when we add in the dependency platform on all projects, this aligns not just all of their first order dependencies, but their transitives too. Further - when we declare it as part of the API, that goes viral. 
```
  implementation(libs.grpc.api)
  api(platform(commonLibs.hypertrace.bom))
```
A project that depends on the above project will be depending on the same platform too and get the same benefit. Let's say some other service depends on this artifact, but it is using an older grpc version. It provides `grpc-netty:1.56.0`. Without the platform, it'd have misaligned versions using the newer grpc-api and the older grpc-netty, which can lead to runtime conflicts.  Because the platform _also_ declares the grpc-netty version, just by depending on this project we keep our versions all in sync.

I hope that helps and didn't just make things more confusing!